src/solc_0.7/GameToken/GameToken.sol

```diff
@@ -0,0 +1,419 @@
+        }
+
+        _setTokenURI(gameId, uri);
+        emit AssetsRemoved(gameId, assetIds, values, to);
+    }
+
+    /// @notice Get the creator of the token type `id`.
+    /// @param id the id of the token to get the creator of.
+    /// @return the creator of the token type `id`.
+    function creatorOf(uint256 id) public view override returns (address) {
+        require(id != uint256(0), "GAME_NEVER_MINTED");
+        address originalCreator = address(id / CREATOR_OFFSET_MULTIPLIER);
+        address newCreator = _creatorship[originalCreator];
+        if (newCreator != address(0)) {
+            return newCreator;
+        }
+        return originalCreator;
+    }
+
+    /// @notice Return the URI of a specific token
+    /// @param gameId The id of the token
+    /// @return uri The URI of the token
+    function tokenURI(uint256 gameId) public view override returns (string memory uri) {
+        require(_ownerOf(gameId) != address(0), "BURNED_OR_NEVER_MINTED");
+        string memory URI = _metaData[gameId];
+        return URI;
+    }
+
+    /// @notice Function to transfer assets from a burnt GAME
+    /// @param from Previous owner of the burnt game
+    /// @param to Address that will receive the assets
+    /// @param gameId Id of the burnt GAME token
+    /// @param assetIds The assets to recover from the burnt GAME
+    /// @param values The amount of each asset to recover
+    function recoverAssets(
+        address from,
+        address to,
+        uint256 gameId,
+        uint256[] memory assetIds,
+        uint256[] memory values
+    ) public override gameManagerOnly() {
+        _recoverAssets(from, to, gameId, assetIds, values);
+    }
+
+    function _setTokenURI(uint256 gameId, string memory URI) internal {
+        _metaData[gameId] = URI;
+    }
+
+    function _destroyGame(
+        address from,
+        address to,
+        uint256 gameId
+    ) internal notToZero(to) notToThis(to) {
+        address owner = _ownerOf(gameId);
+        require(from == owner, "DESTROY_ACCESS_DENIED");
+        delete _metaData[gameId];
+        _creatorship[creatorOf(gameId)] = address(0);
+        _burn(from, owner, gameId);
+    }
+
+    function _recoverAssets(
+        address from,
+        address to,
+        uint256 gameId,
+        uint256[] memory assetIds,
+        uint256[] memory values
+    ) internal notToZero(to) notToThis(to) {
+        _check_withdrawal_authorized(from, gameId);
+        uint256 length = assetIds.length;
+        require(length > 0, "WITHDRAWAL_COMPLETE");
+        require(assetIds.length == values.length, "INVALID_INPUT_LENGTHS");
+
+        for (uint256 i = 0; i < assetIds.length; i++) {
+            uint256 currentValue = _gameAssets[gameId][assetIds[i]];
+            require(currentValue != 0 && values[i] != 0 && values[i] <= currentValue, "INVALID_ASSET_REMOVAL");
+            _gameAssets[gameId][assetIds[i]] = currentValue.sub(values[i]);
```

Comment:
this could be set to zero, no ?

Reply:
Currently, `_recoverAssets` is used by both `destroyAndRecover` (to perform both when gas limit/number of assets allows it) _**and**_ `recoverAssets` (for when using `destroyGame`, followed by `recoverAssets`.
In the first case, using `.sub` will result in 0, but in the second case (with too many assets) `.sub` is needed as there will still be more assets to withdraw/recover.
Make sense?

Reply:
Using `.sub` allows this to be general-purpose and supports both use-cases. If you feel the gas overhead is too high I can refactor.

Reply:
I see, I think we could remove the values paramater as as far erc1155 is concerned there is no gas difference in transferring 1 or 1,000,000 token of the same id

Reply:
fixed.